From 8cee628220e62e8b2e594c0eb2a4a6cd9913a88a Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sat, 8 May 2021 20:20:39 -0400
Subject: [PATCH 1/3] video: Add video_sync_dirty

This is used to work around the slow redraw of a full video console
display.

The `video_sync_dirty` function is *supposed* to be called in situations
where it makes sense for a dynamic user use case. The easiest example
would be in the sibling patch, where it is applied to `tstc` and `getc`
methods of graphical input methods.

This is not supposed to go upstream. This is a workaround for the issue,
which would be better fixed with some kind of co-operative "multi"
tasking, where instead the video-uclass would add a task to sync
"whenever the system can" (if not asked for already).
---
 drivers/video/video-uclass.c | 23 +++++++++++++++++++----
 include/video.h              |  7 +++++++
 2 files changed, 26 insertions(+), 4 deletions(-)

diff --git a/drivers/video/video-uclass.c b/drivers/video/video-uclass.c
index 01e8af5ac67..5ed6fccf89b 100644
--- a/drivers/video/video-uclass.c
+++ b/drivers/video/video-uclass.c
@@ -64,6 +64,8 @@ struct video_uc_priv {
 	ulong video_ptr;
 };
 
+static bool video_is_dirty = false;
+
 void video_set_flush_dcache(struct udevice *dev, bool flush)
 {
 	struct video_priv *priv = dev_get_uclass_priv(dev);
@@ -199,11 +201,17 @@ int video_sync(struct udevice *vid, bool force)
 	 */
 #if defined(CONFIG_ARM) && !CONFIG_IS_ENABLED(SYS_DCACHE_OFF)
 	struct video_priv *priv = dev_get_uclass_priv(vid);
+	static ulong last_sync;
 
-	if (priv->flush_dcache) {
-		flush_dcache_range((ulong)priv->fb,
-				   ALIGN((ulong)priv->fb + priv->fb_size,
-					 CONFIG_SYS_CACHELINE_SIZE));
+	video_is_dirty = true;
+	if (force || get_timer(last_sync) > 10) {
+		if (priv->flush_dcache) {
+			flush_dcache_range((ulong)priv->fb,
+					   ALIGN((ulong)priv->fb + priv->fb_size,
+						 CONFIG_SYS_CACHELINE_SIZE));
+		}
+		video_is_dirty = false;
+		last_sync = get_timer(0);
 	}
 #elif defined(CONFIG_VIDEO_SANDBOX_SDL)
 	struct video_priv *priv = dev_get_uclass_priv(vid);
@@ -247,6 +255,13 @@ bool video_is_active(void)
 	return false;
 }
 
+void video_sync_dirty(void)
+{
+	if (video_is_dirty) {
+		video_sync_all();
+	}
+}
+
 int video_get_xsize(struct udevice *dev)
 {
 	struct video_priv *priv = dev_get_uclass_priv(dev);
diff --git a/include/video.h b/include/video.h
index 43e2c899778..70a1395125f 100644
--- a/include/video.h
+++ b/include/video.h
@@ -179,6 +179,13 @@ int video_sync(struct udevice *vid, bool force);
  */
 void video_sync_all(void);
 
+/**
+ * video_sync_dirty() - Sync all devices' frame buffers when dirty
+ *
+ * This calls video_sync_all() if needed.
+ */
+void video_sync_dirty(void);
+
 /**
  * video_bmp_display() - Display a BMP file
  *
-- 
2.35.1


From 475380e71eba1240d65380e9165c24fe638f9efc Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sat, 8 May 2021 20:35:53 -0400
Subject: [PATCH 2/3] input: sync video on user input request

This "finishes" trailing, or hanging, video updates that might not have
been synced yet. See the previous sibling commit.
---
 common/usb_kbd.c                | 11 +++++++++++
 drivers/input/keyboard-uclass.c | 11 +++++++++++
 2 files changed, 22 insertions(+)

diff --git a/common/usb_kbd.c b/common/usb_kbd.c
index 352d86fb2ec..ea80b62f2e6 100644
--- a/common/usb_kbd.c
+++ b/common/usb_kbd.c
@@ -20,6 +20,7 @@
 #ifdef CONFIG_SANDBOX
 #include <asm/state.h>
 #endif
+#include <video.h>
 
 #include <usb.h>
 
@@ -404,6 +405,11 @@ static int usb_kbd_testc(struct stdio_dev *sdev)
 	 */
 	unsigned long poll_delay = CONFIG_SYS_HZ / 50;
 
+#ifdef CONFIG_DM_VIDEO
+	// For "fast console"
+	video_sync_dirty();
+#endif
+
 #ifdef CONFIG_CMD_NET
 	/*
 	 * If net_busy_flag is 1, NET transfer is running,
@@ -441,6 +447,11 @@ static int usb_kbd_getc(struct stdio_dev *sdev)
 	struct usb_device *usb_kbd_dev;
 	struct usb_kbd_pdata *data;
 
+#ifdef CONFIG_DM_VIDEO
+	// For "fast console"
+	video_sync_dirty();
+#endif
+
 	dev = stdio_get_by_name(sdev->name);
 	usb_kbd_dev = (struct usb_device *)dev->priv;
 	data = usb_kbd_dev->privptr;
diff --git a/drivers/input/keyboard-uclass.c b/drivers/input/keyboard-uclass.c
index aefc8e825e2..caf5b06b215 100644
--- a/drivers/input/keyboard-uclass.c
+++ b/drivers/input/keyboard-uclass.c
@@ -9,6 +9,7 @@
 #include <dm.h>
 #include <keyboard.h>
 #include <log.h>
+#include <video.h>
 
 static int keyboard_start(struct stdio_dev *sdev)
 {
@@ -38,6 +39,11 @@ static int keyboard_tstc(struct stdio_dev *sdev)
 	struct keyboard_priv *priv = dev_get_uclass_priv(dev);
 	struct keyboard_ops *ops = keyboard_get_ops(dev);
 
+#ifdef CONFIG_DM_VIDEO
+	// For "fast console"
+	video_sync_dirty();
+#endif
+
 	/* Just get input to do this for us if we can */
 	if (priv->input.dev)
 		return input_tstc(&priv->input);
@@ -53,6 +59,11 @@ static int keyboard_getc(struct stdio_dev *sdev)
 	struct keyboard_priv *priv = dev_get_uclass_priv(dev);
 	struct keyboard_ops *ops = keyboard_get_ops(dev);
 
+#ifdef CONFIG_DM_VIDEO
+	// For "fast console"
+	video_sync_dirty();
+#endif
+
 	/* Just get input to do this for us if we can */
 	if (priv->input.dev)
 		return input_getc(&priv->input);
-- 
2.35.1


From a2b34d3779f30d3f727c8aad975da3bc17761132 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Mon, 10 May 2021 21:20:51 -0400
Subject: [PATCH 3/3] vidconsole-uclass: Force a sync on newlines

---
 drivers/video/vidconsole-uclass.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/video/vidconsole-uclass.c b/drivers/video/vidconsole-uclass.c
index 95200019363..a46878240f6 100644
--- a/drivers/video/vidconsole-uclass.c
+++ b/drivers/video/vidconsole-uclass.c
@@ -116,7 +116,7 @@ static void vidconsole_newline(struct udevice *dev)
 	}
 	priv->last_ch = 0;
 
-	ret = video_sync(dev->parent, false);
+	ret = video_sync(dev->parent, true);
 	if (ret) {
 #ifdef DEBUG
 		console_puts_select_stderr(true, "[vc err: video_sync]");
@@ -373,7 +373,7 @@ static void vidconsole_escape_char(struct udevice *dev, char ch)
 			int ret;
 
 			video_clear(dev->parent);
-			ret = video_sync(dev->parent, false);
+			ret = video_sync(dev->parent, true);
 			if (ret) {
 #ifdef DEBUG
 				console_puts_select_stderr(true, "[vc err: video_sync]");
-- 
2.35.1

